File: weekly_390th/C.cpp
class Solution {
public:
    vector<long long> mostFrequentIDs(vector<int>& nums, vector<int>& freq) {
        int n = nums.size();
        unordered_map<int, long long> cnt;
        vector<long long> res(n);
        multiset<long long>ms;

        for(int i =0; i < n;i++) {
            auto it = ms.find(cnt[nums[i]]);
            if( it != ms.end()) {
                ms.erase(it);
            }
            cnt[nums[i]] += freq[i];
            ms.insert(cnt[nums[i]]);
            res[i] = *ms.rbegin();
        }
        return res;
    }
};

// 这里的ms.erase(it)很有讲究，it会只删迭代的那一个数字，如果是ms.erase(nums[i]])会直接删除所有符合这个值的element

File: weekly_390th/C.go
// https://pkg.go.dev/github.com/emirpasic/gods/v2@v2.0.0-alpha
func mostFrequentIDs(nums, freq []int) []int64 {
	cnt := map[int]int{}
	t := redblacktree.New[int, int]()
	ans := make([]int64, len(nums))
	for i, x := range nums {
		// 减少一次 cnt[x] 的出现次数
		node := t.GetNode(cnt[x])
		if node != nil {
			node.Value--
			if node.Value == 0 {
				t.Remove(node.Key)
			}
		}

		cnt[x] += freq[i]

		// 增加一次 cnt[x] 的出现次数
		node = t.GetNode(cnt[x])
		if node == nil {
			t.Put(cnt[x], 1)
		} else {
			node.Value++
		}
		ans[i] = int64(t.Right().Key)
	}
	return ans
}

